### PR TITLE
Additional GPU targets for DBCSR

### LINKED ISSUE #28554: cp2k cuda_arch=80 support needed
### Steps to reproduce

concretization failed when setting cuda_arch=80 see https://software.nersc.gov/NERSC/spack-infrastructure/-/jobs/128921. 
```
==> Starting concretization pool with 16 processes
==> Error: cp2k+cuda cuda_arch=80 is unsatisfiable, errors are:
  A conflict was triggered
    To see full clingo unsat cores, re-run with `spack --show-cores=full`
    For full, subset-minimal unsat cores, re-run with `spack --show-cores=minimized
    Warning: This may take (up to) hours for some specs
```

I believe the error is due to fact we dont have a mapping for cuda_arch=80 which is A100

https://github.com/spack/spack/blob/93377942d17d356aaae793c5d665a3e48ab7649e/var/spack/repos/builtin/packages/cp2k/package.py#L532-L538

### Error message

_No response_

### Information on your system

```
(spack-pyenv) e4s:login29> spack debug report
* **Spack:** 0.17.1-930-c38d34223a
* **Python:** 3.6.13
* **Platform:** cray-sles15-zen3
* **Concretizer:** clingo
```

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by RMeli**
Fix #28554?

### CODE REVIEW

File: var/spack/repos/builtin/packages/dbcsr/package.py [hunk 1]
```diff
@@ -75,6 +77,10 @@ class Dbcsr(CMakePackage, CudaPackage, ROCmPackage):
     conflicts("+cuda", when="cuda_arch=none", msg=cuda_msg)
 
     dbcsr_amdgpu_targets = "gfx906"
+    with when("@2.1.0"):
+        dbcsr_amdgpu_targets = (*dbcsr_amdgpu_targets, "gfx908")  # Mi100
```
Comment (RMeli) [inline var/spack/repos/builtin/packages/dbcsr/package.py:81]:
```suggestion
        dbcsr_amdgpu_targets = (dbcsr_amdgpu_targets, "gfx908")  # Mi100
```

File: var/spack/repos/builtin/packages/dbcsr/package.py [hunk 2]
```diff
@@ -75,6 +77,10 @@ class Dbcsr(CMakePackage, CudaPackage, ROCmPackage):
     conflicts("+cuda", when="cuda_arch=none", msg=cuda_msg)
 
     dbcsr_amdgpu_targets = "gfx906"
+    with when("@2.1.0"):
+        dbcsr_amdgpu_targets = (dbcsr_amdgpu_targets, "gfx908")  # Mi100
+    with when("@2.3.0"):
+        dbcsr_amdgpu_targets = (*dbcsr_amdgpu_targets, "gfx90a")  # Mi250
```
Comment (RMeli) [inline var/spack/repos/builtin/packages/dbcsr/package.py]:
```suggestion
    dbcsr_amdgpu_targets = ("gfx906",)
    with when("@2.1.0:"):
        dbcsr_amdgpu_targets = (*dbcsr_amdgpu_targets, "gfx908")  # Mi100
    with when("@2.3.0:"):
        dbcsr_amdgpu_targets = (*dbcsr_amdgpu_targets, "gfx90a")  # Mi250
```

File: var/spack/repos/builtin/packages/dbcsr/package.py [hunk 3]
```diff
@@ -66,6 +66,8 @@ class Dbcsr(CMakePackage, CudaPackage, ROCmPackage):
     # properties, since the parent class defines constraints for different archs
     # Instead just mark all unsupported cuda archs as conflicting.
     dbcsr_cuda_archs = ("35", "37", "60", "70")
+    with when("@2.3.0:"):  # Optimised A100 parameters only for @2.5.0:
+        dbcsr_cuda_archs = (*dbcsr_cuda_archs, "80")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/dbcsr/package.py:70]: This is wrong. The context manager has effect only on directives, not on general Python assignments. To add values that are conditional to a constraint, see [here](https://spack.readthedocs.io/en/latest/packaging_guide.html#conditional-possible-values).

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/dbcsr/package.py (hunk 3 above)
Comment (RMeli) [reply var/spack/repos/builtin/packages/dbcsr/package.py:70]: Thanks for pointing it out @alalazo. It seemed to work on the local tests I did, so I thought it was OK.

If I understand correctly, `conditional` returns a `list` of `Value`s which makes the later `if arch not in dbcsr_cuda_archs` test fail.

**Comment by RMeli** [comment]
Superseded by #38939

**Pull request closed without merging**